File: connect.py
import requests
import logging
import sys
import time
import random

status_url = "http://192.168.9.8/include/auth_action.php?k="
login_url = "http://192.168.9.8/include/auth_action.php"
user_name = "2017221102009"
pass_word = "self1098,"


def init_logging():
    # init logger
    logger = logging.getLogger()
    log_file = sys.path[0] + "/log/" + time.strftime("%Y%m%d%H%M%S", time.localtime()) + ".log"
    formatter = "[%(asctime)s] [%(levelname)s] %(message)s"
    file_handler = logging.FileHandler(log_file, encoding="utf-8")

    logging.basicConfig(level="INFO", format=formatter)
    file_handler.setFormatter(logging.Formatter(formatter))
    logger.addHandler(file_handler)


def check_status():
    keys = random.randint(1, 100000)
    payload = {'action': 'get_online_info', 'key': keys}
    res = requests.post(url=status_url+str(keys), data=payload)
    return not (res.text == "not_online")


def login():
    ac_id = 1
    user_mac = ""
    user_ip = ""
    nas_ip = ""
    save_me = 0
    domain = "@uestc"
    ajax = 1
    payload = {
        "action": "login",
        "username": user_name,
        "password": pass_word,
        "ac_id": ac_id,
        "user_mac": user_mac,
        "user_ip": user_ip,
        "nas_ip": nas_ip,
        "save_me": save_me,
        "domain": domain,
        "ajax": ajax,
    }
    res = requests.post(url=login_url, data=payload)
    if "login_ok" in res.text:
        logging.info("login successfully.")
        return True
    else:
        return False


def check_loop():
    while True:
        if not check_status():
            logging.info("not online, now try to login.")
            if not login():
                logging.error("login failed, wait 20 seconds to reconnect.")
                time.sleep(20)
                continue
            else:
                continue
        else:
            logging.info("network online, sleep for 10 minutes.")
            time.sleep(60*10)


def main():
    init_logging()
    logging.info("app started.")
    if not check_status():
        login()
    check_loop()


if __name__ == "__main__":
    main()

File: connect.go
package main

import (
	"fmt"
	"gopkg.in/ini.v1"
	"io/ioutil"
	"log"
	"math/rand"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"
)

const statusUrl = "http://192.168.9.8/include/auth_action.php?k="
const loginUrl = "http://192.168.9.8/include/auth_action.php"
var userName = "username"
var passWord = "password"
var pidFile = "connect.pid"
var settingFile = "settings.ini"
var checkInterval time.Duration = 600

func main() {
	fmt.Println("Usage:\n    command {setting_file_path} {pid_file_path}")
	// 第一个参数是设置文件的路径
	if len(os.Args) > 1 {
		settingFile = os.Args[1]
	}
	// 第二个参数是 pid 文件的路径
	if len(os.Args) > 2 {
		pidFile = os.Args[2]
	}
	// 第三个是日志文件的路径
	// TODO
	res := loadConfig(settingFile)
	if !res {
		return
	}
	if !writePid(pidFile) {
		log.Println("Failed to write pid file: " + pidFile)
	}
	for {
		isLogin := checkStatus()
		if !isLogin {
			loginStatus := login()
			if !loginStatus {
				log.Println("Login failed, sleep for 1 minute")
				time.Sleep(time.Minute * 1)
			} else {
				log.Printf("Login successfully, sleep for %d seconds\n", checkInterval)
				time.Sleep(time.Second * checkInterval)
			}
			continue
		}
		log.Printf("Network has been online, sleep for %d seconds\n", checkInterval)
		time.Sleep(time.Second * checkInterval)
	}
}

func loadConfig(filename string) bool {
	config, err := ini.Load(filename)
	if err != nil {
		log.Println("Load settings file error: " + err.Error())
		return false
	}
	userName = config.Section("user").Key("username").String()
	passWord = config.Section("user").Key("password").String()
	checkTime, err := config.Section("user").Key("check_interval").Int()
	if err != nil {
		log.Println("Error occurred while parsing check interval")
	} else {
		checkInterval = time.Duration(checkTime)
	}
	log.Printf("Load settings from %s successfully.\n", filename)
	return true
}

func checkStatus() bool {
	keys := rand.Intn(10000)
	urls := statusUrl + strconv.Itoa(keys)
	payload := url.Values{
		"action": {"get_online_info"},
		"keys": {strconv.Itoa(keys)},
	}
	res, err := http.PostForm(urls, payload)
	if err != nil {
		log.Println("Check connection status failed")
		return false
	}
	defer res.Body.Close()
	body, err := ioutil.ReadAll(res.Body)
	text := string(body)
	if text == "not_online" {
		return false
	} else {
		data := strings.Split(text, ",")
		dataBytes, _ := strconv.Atoi(data[0])
		linkTime, _ := strconv.Atoi(data[1])
		log.Println("Account Status: Data Usage: " + strconv.Itoa(dataBytes >> 20) + " MB, Connection Time: " +
			(time.Duration(linkTime) * time.Second).String() + ", IP Address: " + data[5])
		return true
	}
}

func login() bool {
	payload := url.Values{
		"action": {"login"},
		"username": {userName},
		"password": {passWord},
		"ac_id": {"1"},
		"user_mac": {""},
		"user_ip": {""},
		"nas_ip": {""},
		"save_me": {"0"},
		"domain": {"@uestc"},
		"ajax": {"1"},
	}
	res, err := http.PostForm(loginUrl, payload)
	if err != nil {
		log.Println("Login connect failed.")
		return false
	}
	body, err := ioutil.ReadAll(res.Body)
	if err != nil {
		log.Println(err.Error())
		return false
	}
	defer res.Body.Close()
	text := string(body)
	if strings.Contains(text, "login_ok") {
		log.Println("Login successfully.")
		return true
	}
	log.Println("Login Failed.")
	return false
}

func writePid(filename string) bool {
	pid := strconv.Itoa(os.Getpid())
	file, err := os.Create(filename)
	if err != nil {
		return false
	}
	_, _ = file.WriteString(pid)
	defer file.Close()
	log.Printf("Write to pid file(%s) successfully, pid: %s", filename, pid)
	return true
}
